icu4j/main/core/src/main/java/com/ibm/icu/message2/InputSource.java

```diff
@@ -62,10 +62,22 @@ int readCodePoint() {
         return c;
     }
 
-    // Backup a number of characters.
+    // Backup a number of code points.
     void backup(int amount) {
```

Comment:
This method works on code units (Java `char`), not code points.

That's because `int readCodePoint()` also works on code units.
It returns a code point, but the current offset in the input (`cursor`) is expresses in code units.
So when it finds a code point above BMP it advances with 2

Reasons:
 - When one wants to parse something from an offset, it is faster to express the offset in the same kind of units we use for storage. We store using `char`, offset in code units. Otherwise we must iterate all the way to the offset in code points. Our input buffer uses `char`
 - Easier to deal with not-properly paired surrogates. The spec does not try to enforce the surrogate correctness, and I don't think it is the job of the MF2 to do that. If I pass `".....\uDC00...{$user}!` there is no reason to not format to `".....\uDC00...John!` 

Even if there is disagreement about the reasons, and we decide to express offsets in code points, this fix is not sufficient.

We would also have to change `getPosition()`, `skip(int amount)`, `gotoPosition(int position)`
Because we either work on code points, or on code units, we should not mix and match.

Reply:
Thanks for the explanation. The issue isn't really with `backup()`, but how it's used in conjunction with `readCodePoint()` (reading a code point that happens to be a wide char, and then calling `backup(1)` to "push it back", is a bug).

In [eee547c](https://github.com/unicode-org/icu/pull/3092/commits/eee547c25ba7275947dde0a22a4d34ab10a19974) I reverted the change to `backup()` and fixed the three places I could find where `readCodePoint()` and `backup()` were being used in that way.